### PR TITLE
Framework for exporter types and options

### PR DESCRIPTION
**Pull request by nyalldawson**
* Not to be merged *

I'm opening this PR for feedback. This sets the way for allowing non-folder based exports, such as uploading direct to an FTP site.

It adds a new base class for individual Exporter objects, which handle the export of a web map.

An ExporterRegistry tracks all available exporter types and interaction with the current project (storage and retrieval of settings).

Currently only the existing folder export and a non-functional FTP upload exporter are available. The folder export option should all be functioning!

This framework should make it nice and easy to add additional exporters as required.

### CODE REVIEW

**Comment by nyalldawson** [comment]
@tomchadwin what process do you use to run the tests locally?

**Comment by tomchadwin** [comment]
@nyalldawson re: local tests. I'm afraid I don't run them locally, as I'm on Windows, so I rely entirely on Travis. @jnicho02 did some work at FOSS4GUK last year to get the tests running on Mac, and I've run them successfully on Ubuntu, so let me have a look and get back to you. There is something in CONTRIBUTING.md, but I've not tried it out for a while.

**Comment by tomchadwin** [comment]
OK, specifically, the local testing instructions are here:

https://github.com/tomchadwin/qgis2web/blob/master/CONTRIBUTING.md#testing

However, I can't get them to work, presumably because I am getting my QGIS install path wrong. I hope this is just down to my Linux/QGIS ignorance, and you'll be able to get it working easily. Let me know.

The other person who knows about this is @akbargumbira, so if you are around, Akbar, Nyall and I would appreciate a little guidance.

**Comment by nyalldawson** [comment]
OK - this is functional now (albeit with a basic UI - eg there's no feedback during the upload or if things go wrong!). 

But I'm happy with the design and think this is ready for review and possible merge now.

File: maindialog.py [hunk 1]
```diff
@@ -153,13 +157,11 @@ def toggleOptions(self):
                         treeOption.setDisabled(False)
 
     def previewMap(self):
-        if not webkit_available:
```
Comment (tomchadwin) [inline maindialog.py]: This was added to handle Xenial installs which had no QWebKit - probably needs to be tested in that environment

File: test/test_qgis2web_dialog.py [hunk 2]
```diff
@@ -23,6 +23,7 @@
 from qgis.core import QgsVectorLayer, QgsMapLayerRegistry, QgsCoordinateReferenceSystem
 from PyQt4 import QtCore, QtTest
 from PyQt4.QtCore import *
+from osgeo import gdal
```
Comment (tomchadwin) [inline test/test_qgis2web_dialog.py]: Is upgrading GDAL in the Travis setup not an alternative option?

**Comment by tomchadwin** [comment]
This looks fantastic. Just going to test locally, and then we'll definitely merge. Did you get local tests working?

File: maindialog.py (hunk 1 above)
Comment (nyalldawson) [inline maindialog.py]: That's what I'm running under :) That check has been moved to loadPreviewFile to help with the test suite - so basically now the behavior for with/without webkit is the same right up until the actual preview is shown.

File: test/test_qgis2web_dialog.py (hunk 2 above)
Comment (nyalldawson) [inline test/test_qgis2web_dialog.py]: It's tricky. I think a better approach here might be to have a set of acceptable control files, and if the test matches either then it passes. FYI the difference is that on travis coordinates like 18.00 will be truncated to 18.0, but on newer gdal they will get fully truncated to just '18'

**Comment by nyalldawson** [comment]
> Did you get local tests working?

Yes - running 'make test' runs them all without issue now for me. There's a bit of environment setup required (eg setting the correct QGIS_PREFIX_PATH) which must be done first, but that's quite specific to local installs and I don't believe there's anything we can do to help here (maybe just document it...)

One thing I'm not sure on - I don't think coveralls is checking the new exporter.py file. Do you know what's required to add this?

> Obviously some additional functionality required (error trapping, open remote files in browser after export, not local ones), but I think you're right that this is good to merge now.

Agreed. I think a feedback window showing each file which is uploaded, and with a cancel button, is required. I just need to wait for confirmation from my client and then I'll add these extra polish.

**Comment by tomchadwin** [comment]
> I don't think coveralls is checking the new exporter.py file

Yes, I'll sort that.

Two other things I think need consideration:

1. Creation of the remote folder/path of not already present. This removes the need to prepare the ground with a separate FTP client. 

2. Allow overwriting of remote files (unless it just happened to be my FTP account permissions which stopped this from working). Otherwise you can't re-export. 

I think both are required for the most common pattern of usage.

**Pull request merged**